[[ Bug 22763 ]] Fix resizing mobilePickPhoto for API Level < 24

livecodepanos opened the pull request:
This patch ensures that calling `mobilePickPhoto` with width and height params works in older Android versions too. Previously, a constructor for `ExifInterface` was used, that was available in API 24+ (Android 7+).

This patch uses another constructor, that is available in older APIs as well.

Closes https://quality.livecode.com/show_bug.cgi?id=22763

runrevmark commented on engine/src/java/com/runrev/android/Engine.java:
```diff
@@ -2232,9 +2232,7 @@ else if (resultCode == Activity.RESULT_OK)
-					InputStream t_exif_in = ((LiveCodeActivity)getContext()).getContentResolver().openInputStream(t_photo_uri);
-					
-					ExifInterface t_exif = new ExifInterface(t_exif_in);
+					ExifInterface t_exif = new ExifInterface(t_photo_uri.getPath());
```
I'm a little concerned that this will break on newer Android versions. Android versions >= 7 (IIRC - API Level 24 by any chance - that's where the new constructor was introduced) are more strict about how apps access files.

In this case the input could be either a Uri from picking from a library, or presumably a temp file from taking a photo.

Can you confirm that all aspects of mobilePickPhoto still work with this change in both Andoid6 and 7+?

runrevmark reviewed (changes requested):
I've updated the title of the PR, could you make the main commit message title match it please :)

runrevmark commented on engine/src/java/com/runrev/android/Engine.java:
```diff
@@ -2231,10 +2231,16 @@ else if (resultCode == Activity.RESULT_OK)
-					
-					ExifInterface t_exif = new ExifInterface(t_exif_in);
+                    
+                    ExifInterface t_exif;
```
As the rest of this file is indent with tabs, make sure the changes are

There appears to already be an InputStream created for the photo above (`t_in`) so we can resue that here. This means that this code can be a little more compact. I suggest:
```
/* In API Level >= 24, you have to use an input stream to make sure that access
 * to a photo in the library is granted. Before that you can just open the photo's
 * file direct. */
ExifInterface t_exif;
if (Build.VERSION.SDK_INT >= 24)
{
    t_exit = new ExifInterface(t_in);
}
else
{
    t_exif = new ExifInterface(t_photo_uri.getPath());
}
```

livecodepanos committed:
[[ Bug 22763 ]] Fix resizing mobilePickPhoto for API Level < 24

This patch ensures that calling mobilePickPhoto with width and height
params works in older Android versions too. Previously, a constructor
for "ExifInterface" was used, that was available in API 24+ (Android 7+).
This patch checks the Android version of the device and uses a different
constructor on Android < 7.

runrevmark commented:
@livecode-vulcan review ok 7dafc03

livecode-vulcan commented:
💙 review by @runrevmark ok 7dafc033893bcefbd0e2653029f3b021a201d46d

livecode-vulcan commented:
😎 test success 7dafc033893bcefbd0e2653029f3b021a201d46d

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/836)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/245)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/245)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/245)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/245)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2362)
* try-community-universal-ios-iphoneos13.5: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos13.5/builds/13)
* try-community-universal-ios-iphonesimulator13.5: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator13.5/builds/13)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2390)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2378)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2377)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2412)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2352)

livecodepanos commented on engine/src/java/com/runrev/android/Engine.java:
```diff
@@ -2232,9 +2232,18 @@ else if (resultCode == Activity.RESULT_OK)
-					InputStream t_exif_in = ((LiveCodeActivity)getContext()).getContentResolver().openInputStream(t_photo_uri);
-					
-					ExifInterface t_exif = new ExifInterface(t_exif_in);
+					/* In API Level >= 24, you have to use an input stream to make sure that access
```
Hmm, using the existing InputStream `t_in` causes the captured pic to be rotated 90 degrees. I am wondering if this was the reason @montegoulding used a new one (i.e. the `t_exif_in` one) in the first place.

runrevmark replied:
Oh! Yes - the reason is that we've used the t_in stream above! Oops!

So below we should do:
```
if (Build.VERSION....)
{
    t_in.close();
    t_in = ((LiveCodeActivity)getContext()).getContentResolver().openInputStream(t_photo_uri);
}
```
This will close the (now finished) original stream, and then create a new one.
